### PR TITLE
feat: add support for uuid format

### PR DESCRIPTION
**Pull request by gazoakley**
Closes https://github.com/Redocly/redoc/issues/1246

Adds support for generating example UUIDs. To make the examples more realistic (and ensure consistency between examples) the property name is used when available to deterministically generate the UUID. Given the following schema:

```yaml
type: object
properties:
  fooId:
    type: string
    format: uuid
  barId:
    type: string
    format: uuid
```

The following example will be generated:

```json
{
  "fooId": "1e6c707e-f129-5967-b5d5-91b8af7aaf12",
  "barId": "9c7c400b-3b79-57c7-8eef-be9496e711c9"
}

### CODE REVIEW

**Comment by gazoakley** [comment]
@RomanHotsiy: The Travis build appears to be failing when trying to launch xvfb to run tests. They do however pass locally:

```
Integration
    Primitives
      ✓ should sample string
      ✓ should sample number
      ✓ should sample boolean
      ✓ should use default property
      ✓ should use null if type is not specified
    Objects
      ✓ should sample object without properties
      ✓ should sample object with property
      ✓ should sample object with property with default value
      ✓ should sample object with more than one property
      ✓ should sample both properties and additionalProperties
    AllOf
      ✓ should sample schema with allOf
      ✓ should throw for schemas with allOf with different types
      ✓ deep array
      ✓ should not be confused by subschema without type
OpenAPI Sampler: found allOf with "array" type. Result may be incorrect
      ✓ should not throw for array allOf
      ✓ should sample schema with allOf even if some type is not specified
      ✓ should merge deep properties
    Example
      ✓ should use example
      ✓ should use falsy example
      ✓ should use enum
    Detection
      ✓ should detect autodetect types based on props
    oneOf and anyOf
      ✓ should support oneOf
      ✓ should support anyOf
oneOf and anyOf are not supported on the same level. Skipping anyOf
      ✓ should prefer oneOf if anyOf and oneOf are on the same level 
    $refs
      ✓ should follow $ref
      ✓ should not follow circular $ref
      ✓ should not follow circular $ref if more than one in properties
      ✓ should throw if schema has $ref and spec is not provided
      ✓ should ignore readOnly params if referenced

  sampleArray
    ✓ should return empty array by default
    ✓ should return elements of correct type
    ✓ should return correct number of elements based on minItems
    ✓ should correcly sample tuples

  sampleNumber
    ✓ should return 0 by default
    ✓ should return 0 by default
    ✓ should return minimum if minimum is specified
    ✓ should return minimum +1 for exclusiveMinimum
    ✓ should return minimum +1 for exclusiveMinimum
    ✓ should return maximum if maximum is negative
    ✓ should return maximum -1 if maximum is negative and exclusiveMaximum
    ✓ should return minimum if both minimum and maximum are specified
    ✓ should return middle point if integer is not possible
    ✓ should return closer to minimum possible int

  sampleObject
    ✓ should return emtpy object by default
    ✓ should instantiate all properties
    ✓ should skip readonly properties if skipReadOnly=true
    ✓ should skip readonly properties in nested objects if skipReadOnly=true
    ✓ should skip writeonly properties if writeonly=true
    ✓ should skip writeonly properties in nested objects if writeonly=true
    ✓ should should instantiate 2 additionalProperties
    ✓ should skip non-required properties if skipNonRequired=true
    ✓ should pass propertyName context to samplers

  sampleString
    ✓ should return "string" by default
    ✓ should return string of appropriate length if minLength specified
    ✓ should return string of appropriate length if maxLength specified
    ✓ should return email for format email
    ✓ should return email for format email
    ✓ should return password for format password
    ✓ should return password of appropriate length if minLength specified
    ✓ should return date string for format date
    ✓ should return date string for format date
    ✓ should throw if incorrect maxLength applied to date-time
    ✓ should throw if incorrect minLength applied to date-time
    ✓ should return ip for ipv4 format
    ✓ should return ipv6 for ipv6 format
    ✓ should return vaild hostname for hostname format
    ✓ should return vaild URI for uri format
    ✓ should return valid uuid for uuid format without propertyName context
    ✓ should return valid uuid for uuid format with propertyName context


  69 passing (50ms)

----------|----------|----------|----------|----------|----------------|
File      |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
----------|----------|----------|----------|----------|----------------|
----------|----------|----------|----------|----------|----------------|
All files |      100 |      100 |      100 |      100 |                |
----------|----------|----------|----------|----------|----------------|


=============================== Coverage summary ===============================
Statements   : 100% ( 0/0 )
Branches     : 100% ( 0/0 )
Functions    : 100% ( 0/0 )
Lines        : 100% ( 0/0 )
================================================================================
```

**Comment by gazoakley** [comment]
Hi @romanhotsiy

What about https://github.com/velocityzen/uuid5/blob/master/index.js ? It's only 16 LOC, but would still allow deterministic uuids per property.

**Comment by gazoakley** [comment]
Actually, using `uuid5` means including a lot of dependencies when browserified. I've swapped in a generator based on the code you linked to but made it deterministic.

**Comment by RomanHotsiy** [comment]
@gazoakley thanks for updates. I will review it this week.

File: src/samplers/string.js [hunk 1]
```diff
@@ -60,6 +60,10 @@ function uriSample() {
   return 'http://example.com';
 }
 
+function uuidSample(min, max, propertyName) {
```
Comment (RomanHotsiy) [inline src/samplers/string.js:63]: Underscore unused variables

```suggestion
function uuidSample(_min, _max, propertyName) {
```

File: src/utils.js [hunk 2]
```diff
@@ -46,3 +46,36 @@ export function mergeDeep(...objects) {
     return prev;
   }, Array.isArray(objects[objects.length - 1]) ? [] : {});
 }
+
```
Comment (RomanHotsiy) [inline src/utils.js:49]:
```suggestion

// deterministic UUID sampler
```

File: src/samplers/string.js (hunk 1 above)
Comment (gazoakley) [reply src/samplers/string.js]: Updated: https://github.com/Redocly/openapi-sampler/pull/106/files#diff-bd2268fad45114c88f3962f59318158fR63

File: src/utils.js (hunk 2 above)
Comment (gazoakley) [reply src/utils.js:49]: Updated: https://github.com/Redocly/openapi-sampler/pull/106/files#diff-2b4ca49d4bb0a774c4d4c1672d7aa781R50

**Comment by RomanHotsiy** [comment]
@gazoakley could you also rebase on master so Travis checks run successfully (I fixed it on master)

**Comment by RomanHotsiy** [comment]
Thanks!

**Pull request merged**